frontend/dockerfile/dockerfile2llb/convert.go

```diff
@@ -1001,7 +1002,25 @@ func dispatchCopy(d *dispatchState, cfg copyConfig) error {
 
 	for _, src := range cfg.params.SourcePaths {
 		commitMessage.WriteString(" " + src)
-		if strings.HasPrefix(src, "http://") || strings.HasPrefix(src, "https://") {
+		if gitutil.IsGitRef(src) {
+			if !cfg.isAddCommand {
+				return errors.New("source can't be a git ref for COPY")
+			}
+			g, err := gitutil.ParseGitRef(src)
+			if err != nil {
+				return errors.Wrapf(err, "failed to parse a git ref %q", src)
+			}
+			st := llb.Git(g.Repo, g.Commit, llb.KeepGitDir())
```

Comment:
> I guess we could use a build-arg for this as well?

Should this be rather an flag for `ADD` instruction? e.g., `ADD --keep-git-dir=(true|false) ...`.

Reply:
I'm not sure. Don't like to add lots of very specific flags either.

Reply:
While the git dir is not preserved for build contexts, I'd prefer to keep the git dir for the `ADD` instruction, as Makefiles often need `git rev-parse`, `git describe`, etc.

https://github.com/containerd/containerd/blob/8e74f2af0096a046ac004dc954105b010b9ee2d5/Makefile#L33

Reply:
Could we still do something for this that would allow have same defaults as context but opt-in to keep-git-dir. Maybe a convention on the URL, or a `ARG` definition before the command that controls this(or maybe a flag is fine)?

Reply:
Not a huge fan of complicating the URL microformat.
I also fear that adding an `ARG` like `ARG KEEPGITDIR=true` may conflict with user-defined `ARG` arguments.

Maybe `ADD --keep-git-dir=(true|false) ...` is the best?

Reply:
@tonistiigi SGTY? ↑

Reply:
I think it is fine yes. In addition to plain `keep-git` we might need a way to configure the depth size as well in the future.